pkg/app/piped/planner/kubernetes/kubernetes.go

```diff
@@ -186,39 +187,45 @@ func (p *Planner) Plan(ctx context.Context, in planner.Input) (out planner.Outpu
 
 // First up, checks to see if the workload's `spec.template` has been changed,
 // and then checks if the configmap/secret's data.
-func decideStrategy(olds, news []provider.Manifest) (progressive bool, desc string) {
-	oldWorkload, ok := findWorkload(olds)
-	if !ok {
+func decideStrategy(olds, news []provider.Manifest, workloadRefs []config.K8sResourceReference) (progressive bool, desc string) {
+	oldWorkloads := findWorkloadManifests(olds, workloadRefs)
+	if len(oldWorkloads) == 0 {
 		desc = "Quick sync by applying all manifests because it was unable to find the currently running workloads"
 		return
 	}
-	newWorkload, ok := findWorkload(news)
-	if !ok {
+	newWorkloads := findWorkloadManifests(news, workloadRefs)
+	if len(newWorkloads) == 0 {
 		desc = "Quick sync by applying all manifests because it was unable to find workloads in the new manifests"
 		return
 	}
 
-	// If the workload's pod template was touched
-	// do progressive deployment with the specified pipeline.
-	diffResult, err := provider.Diff(oldWorkload, newWorkload)
-	if err != nil {
-		progressive = true
-		desc = fmt.Sprintf("Sync progressively due to an error while calculating the diff (%v)", err)
-		return
-	}
-	diffNodes := diffResult.Nodes()
+	workloads := findUpdatedWorkloads(oldWorkloads, newWorkloads)
+	diffs := make(map[provider.ResourceKey]diff.Nodes, len(workloads))
 
-	templateDiffs := diffNodes.FindByPrefix("spec.template")
-	if len(templateDiffs) > 0 {
-		progressive = true
-
-		if msg, changed := checkImageChange(templateDiffs); changed {
-			desc = msg
+	for _, w := range workloads {
+		// If the workload's pod template was touched
+		// do progressive deployment with the specified pipeline.
+		diffResult, err := provider.Diff(w.old, w.new)
+		if err != nil {
+			progressive = true
+			desc = fmt.Sprintf("Sync progressively due to an error while calculating the diff (%v)", err)
 			return
 		}
+		diffNodes := diffResult.Nodes()
+		diffs[w.new.Key] = diffNodes
 
-		desc = fmt.Sprintf("Sync progressively because pod template of workload %s was changed", newWorkload.Key.Name)
-		return
+		templateDiffs := diffNodes.FindByPrefix("spec.template")
+		if len(templateDiffs) > 0 {
+			progressive = true
+
+			if msg, changed := checkImageChange(templateDiffs); changed {
+				desc = msg
+				return
+			}
+
+			desc = fmt.Sprintf("Sync progressively because pod template of workload %s was changed", w.new.Key.Name)
+			return
```

Comment:
nits, move `progressive = true` to L225 before return statement (along with `desc = ...` ) make this more readable 👀

Reply:
In that case, it doesn't sync progressively when the image was changed, doesn't it?

Reply:
This effort is for making the description as readable as possible, you know?

Reply:
oh, I missed it 👍 But to me maybe `progressive = true` should be added before line L222 too 🤔 But it's not strong opinion 🙆‍♂️